File: src/main/java/edu/cnm/deepdive/sieve/array/Sieve.java
package edu.cnm.deepdive.sieve.array;

import java.util.Arrays;
import java.util.stream.IntStream;

public class Sieve {

  private static int[] sieve(int limit) {
    boolean[] candidates = new boolean[limit + 1];
    Arrays.fill(candidates, 2, limit + 1, true);
    for (int prime = 2; prime <= Math.sqrt(limit); prime++) {
      if (candidates[prime] == true) {
        for (int multiple = prime * prime; multiple <= limit; multiple += prime) {
          candidates[multiple] = false;
        }
      }
    }
    return IntStream
        .range(0, candidates.length)
        .filter((value) -> candidates[value])
        .toArray();
  }

  public static void main(String... args) {
    int upperBound = 1_000_000;
    long start = System.currentTimeMillis();
    int[] primes = sieve(upperBound);
    long end = System.currentTimeMillis();
    System.out.printf("Java Sieve with boolean[] and int[]:%n"
            + "%1$,d primes found between %2$,d and %3$,d in %4$d ms.%n",
        primes.length, primes[0], primes[primes.length - 1], end - start);
  }

}

File: src/main/kotlin/edu/cnm/deepdive/sieve/bitset/sieve.kt
package edu.cnm.deepdive.sieve.bitset

import java.util.BitSet
import kotlin.math.sqrt

fun sieve(limit: Int): BitSet {
    val candidates = BitSet(limit + 1)
    candidates.set(2, limit + 1)
    var prime = candidates.nextSetBit(0)
    while (prime <= sqrt(limit.toDouble())) {
        for (multiple in (prime * prime)..limit step prime) {
            candidates.clear(multiple)
        }
        prime = candidates.nextSetBit(prime + 1)
    }
    return candidates
}

fun main(args: Array<String>) {
    val start = System.currentTimeMillis()
    val upperBound = 1_000_000
    val primes = sieve(upperBound)
    val end = System.currentTimeMillis()
    println(
        """
            Kotlin Sieve with BitSet: 
            ${primes.cardinality()} primes found between ${primes.nextSetBit(0)} and ${
            primes.previousSetBit(
                primes.size() - 1
            )
        } (inclusive) in ${end - start} ms.
        """.trimIndent()
    )
}